### PR TITLE
Serialization refactoring

### PR DESCRIPTION
**Pull request by vkagamlyk**
*Issue #, if available:*
https://bitquill.atlassian.net/browse/AN-1122

*Description of changes:*
performance enhancements for Golang Driver

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/codecov.yml [hunk 1]
```diff
@@ -16,4 +16,14 @@
...
+coverage:
+  status:
+    project:
+      default:
+        target: auto
+        threshold: 5%
+    patch:
+      default:
+        target: 0%
+
```
Comment (lyndonbauto) [inline .github/codecov.yml]: Nit add newline.

**Comment by lyndonbauto** [comment]
Overall LGTM, going to hold merging for a day or so because this will likely have 3.6-dev and master conflicts as they have additional types.

File: .github/codecov.yml (hunk 1 above)
Comment (simonz-bq) [reply .github/codecov.yml]: That was odd. It did have a newline but I guess the yml format neeeds 2 new lines in order for it to count?

Git warned me about it until I added two, so now there are two.

**Comment by vkagamlyk** [comment]
moving to 3.5-dev

**Pull request closed without merging**